src/vm/disassembler.cpp

```diff
@@ -0,0 +1,308 @@
+        FastInterlockExchangePointer(&s_availableExternalDisassembler, static_cast<ExternalDisassembler *>(nullptr));
+    if (externalDisassembler == nullptr)
+    {
+    #if USE_COREDISTOOLS_DISASSEMBLER
+        // First parameter:
+        // - Empty string for the current architecture
+        // - A string of the form "x86_64-pc-win32"
+        externalDisassembler = External_InitDisasm(Target_Host);
+    #elif USE_MSVC_DISASSEMBLER
+    #ifdef _TARGET_X86_
+        externalDisassembler = ExternalDisassembler::PdisNew(ExternalDisassembler::distX86);
+    #elif defined(_TARGET_AMD64_)
+        externalDisassembler = ExternalDisassembler::PdisNew(ExternalDisassembler::distX8664);
+    #endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+    #endif // USE_COREDISTOOLS_DISASSEMBLER || USE_MSVC_DISASSEMBLER
+    }
+
+    _ASSERTE(externalDisassembler != nullptr);
+    m_externalDisassembler = externalDisassembler;
+}
+
+Disassembler::~Disassembler()
+{
+    LIMITED_METHOD_CONTRACT;
+    _ASSERTE(IsAvailable());
+
+    // Save the external disassembler for future use. We only save one instance, so delete a previously saved one.
+    ExternalDisassembler *externalDisassemblerToDelete =
+        FastInterlockExchangePointer(&s_availableExternalDisassembler, m_externalDisassembler);
+    if (externalDisassemblerToDelete == nullptr)
+    {
+        return;
+    }
+
+#if USE_COREDISTOOLS_DISASSEMBLER
+    External_FinishDisasm(externalDisassemblerToDelete);
+#elif USE_MSVC_DISASSEMBLER
+    delete externalDisassemblerToDelete;
+#endif // USE_COREDISTOOLS_DISASSEMBLER || USE_MSVC_DISASSEMBLER
+}
+
+SIZE_T Disassembler::DisassembleInstruction(const UINT8 *code, SIZE_T codeLength, InstructionType *instructionTypeRef) const
+{
+    LIMITED_METHOD_CONTRACT;
+    _ASSERTE(IsAvailable());
+
+#if USE_COREDISTOOLS_DISASSEMBLER
+    SIZE_T instructionLength = External_DisasmInstruction(m_externalDisassembler, reinterpret_cast<SIZE_T>(code), code, codeLength, false /* PrintAssembly */);
+#elif USE_MSVC_DISASSEMBLER
+    SIZE_T instructionLength =
+        m_externalDisassembler->CbDisassemble(reinterpret_cast<ExternalDisassembler::ADDR>(code), code, codeLength);
+#endif // USE_COREDISTOOLS_DISASSEMBLER || USE_MSVC_DISASSEMBLER
+    _ASSERTE(instructionLength <= codeLength);
+
+    if (instructionTypeRef != nullptr)
+    {
+        if (instructionLength == 0)
+        {
+            *instructionTypeRef = InstructionType::Unknown;
+        }
+        else
+        {
+        #if USE_COREDISTOOLS_DISASSEMBLER
+            *instructionTypeRef = DetermineInstructionType(code, instructionLength);
+        #elif USE_MSVC_DISASSEMBLER
+            *instructionTypeRef = DetermineInstructionType(m_externalDisassembler->Trmt());
+        #endif // USE_COREDISTOOLS_DISASSEMBLER || USE_MSVC_DISASSEMBLER
+        }
+    }
+
+    return instructionLength;
+}
+
+// static
+InstructionType Disassembler::DetermineInstructionType(
+#if USE_COREDISTOOLS_DISASSEMBLER
```

Comment:
It is fine for now but in the future we should get rid of all these "#if USE_COREDISTOOLS_DISASSEMBLER" by separating the implementation into two different files/classes.